core/silkworm/common/endian.hpp

```diff
@@ -171,19 +171,36 @@ Bytes to_big_compact(const uint64_t value);
 //! \remarks A "compact" big endian form strips leftmost bytes valued to zero
 Bytes to_big_compact(const intx::uint256& value);
 
-//! \brief Parses uint64_t from a compacted big endian byte form
-//! \param [in] data : byte view of memory allocation for compacted value. Length must be <= sizeof(uint64_t)
-//! \param [in] allow_leading_zeros : when false, return std::nullopt if data starts with a 0 byte (not compact)
-//! \return A uint64_t with native endianness; std::nullopt if data is invalid
+//! \brief Parses unsigned integer from a compacted big endian byte form
+//! \param [in] data : byte view of compacted value. Length must be <= sizeof(UnsignedInteger)
+//! \param [in] allow_leading_zeros : when false, return std::nullopt if data starts with a 0 byte (i.e. not compact)
+//! \return The corresponding integer with native endianness; std::nullopt if data is invalid
 //! \remarks A "compact" big endian form strips leftmost bytes valued to zero
-std::optional<uint64_t> from_big_compact_u64(const ByteView& data, bool allow_leading_zeros = false);
+template <typename UnsignedInteger>
+static std::optional<UnsignedInteger> from_big_compact(ByteView data, bool allow_leading_zeros = false) {
+    if (data.length() > sizeof(UnsignedInteger)) {
+        return std::nullopt;
+    }
 
-//! \brief Parses uint256 from a compacted big endian byte form
-//! \param [in] data : byte view of memory allocation for compacted value. Length must be <= sizeof(uint256)
-//! \param [in] allow_leading_zeros : when false, return std::nullopt if data starts with a 0 byte (not compact)
-//! \return A uint256 with native endianness; std::nullopt if data is invalid
-//! \remarks A "compact" big endian form strips leftmost bytes valued to zero
-std::optional<intx::uint256> from_big_compact_u256(const ByteView& data, bool allow_leading_zeros = false);
+    UnsignedInteger x{0};
+
+    if (data.empty()) {
+        return x;
+    }
+
+    if (data[0] == 0 && !allow_leading_zeros) {
+        return std::nullopt;
+    }
+
+    auto* ptr{reinterpret_cast<uint8_t*>(&x)};
+    std::memcpy(ptr + (sizeof(UnsignedInteger) - data.length()), &data[0], data.length());
```

Comment:
If allow_leadind_zeros is true, then I think this memcpy would be incorrect as the `ptr + (sizeof(UnsignedInteger) - data.length())` assumes that all the bytes from `data` are significant, right?

Reply:
It should be correct even in that case. Let's say that `UnsignedInteger` is `uint32_t` and the data is from_hex("00AB"). Then after `memcpy(ptr + 2, &data[0], 2)` `x` will have bytes 0, 0, 0, 0xAB, which is the desired big-endian number. But I'll add a test case just to be on the safe side.